bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,80 +1,302 @@
 
-    pub async fn is_healthy(&mut self) -> bool {
-        match self
-            .health_check_client
-            .check(HealthCheckRequest { service: "".to_string() })
-            .await
-        {
-            Ok(response) => response.into_inner().status() == ServingStatus::Serving,
-            Err(err) => {
-                error!("Failed to check worker health ({}): {}", self.address(), err);
-                false
+    // PRIVATE HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns whether the worker should do a health check.
+    ///
+    /// A worker should do a health check if it is healthy or if the time since the first failure
+    /// is greater than the time since the first failure power of 2.
+    ///
+    /// The maximum exponent is [MAX_BACKOFF_EXPONENT], which corresponds to a backoff of
+    /// 2^[MAX_BACKOFF_EXPONENT] seconds.
+    fn should_do_health_check(&self) -> bool {
+        match self.health_status {
+            WorkerHealthStatus::Healthy => true,
+            WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts,
+                first_fail_timestamp,
+                reason: _,
+            } => {
+                let time_since_first_failure = Instant::now() - first_fail_timestamp;
+                time_since_first_failure
+                    > Duration::from_secs(
+                        2u64.pow(failed_attempts.min(MAX_BACKOFF_EXPONENT) as u32),
+                    )
             },
+            WorkerHealthStatus::Unknown => true,
         }
     }
 
-    pub fn is_available(&self) -> bool {
-        self.is_available
+    /// Sets the health status of the worker.
+    fn set_health_status(&mut self, health_status: WorkerHealthStatus) {
+        let previous_health_status = self.health_status.clone();
+        self.health_status = health_status;
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => {
+                if matches!(previous_health_status, WorkerHealthStatus::Unhealthy { .. }) {
+                    self.is_available = true;
+                }
+            },
+            WorkerHealthStatus::Unhealthy { .. } => {
+                WORKER_UNHEALTHY.with_label_values(&[&self.address()]).inc();
+                self.is_available = false;
+            },
+            WorkerHealthStatus::Unknown => {
+                if matches!(previous_health_status, WorkerHealthStatus::Unhealthy { .. }) {
+                    self.is_available = true;
+                }
+            },
+        }
     }
 
-    pub fn set_availability(&mut self, is_available: bool) {
-        self.is_available = is_available;
+    /// Returns whether the worker version is valid.
+    ///
+    /// The version is valid if it is a semantic version and is greater than or equal to the
+    /// current version. We dont check the patch version.
+    fn is_valid_version(&self, version: &str) -> bool {
```

Comment:
This doesn't seem to use `self`, so maybe it could be a separate function. Maybe it could also be a function that compares 2 separate versions, to make it easier to add a couple of tests to make sure it works as expected.